ppafm/cpp/ProbeParticle.cpp

```diff
@@ -341,6 +345,118 @@ DLLEXPORT void getInPoints_LJ( int npoints, double * points_, double * FEs, int
     }
 }
 
+
+DLLEXPORT void computeD3Coeffs(
+    const int natoms_, const double *rs, const int *elems, const double *r_cov, const double *r_cut, const double *ref_cn,
+    const double *ref_c6, const double *r4r2, const double *k, const double *params, const int elem_pp, double *d3_coeffs
+) {
+
+    natoms = natoms_;
+    Ratoms = (Vec3d*)rs;
+
+    const double k1  = k[0];
+    const double k2  = k[1];
+    const double k3  = -k[2];
+    const double k12 = -k1 * k2;
+
+    const double s6 = params[0];
+    const double s8 = params[1];
+    const double a1 = params[2];
+    const double a2 = params[3];
+
+    float L[MAX_REF_CN * MAX_REF_CN];
+
+    // Compute reference C6 weights for the probe particle
+    int pp_ind = elem_pp - 1;
+    double L_pp[MAX_REF_CN];
+    int i;
+    for (i = 0; i < MAX_REF_CN; i++) {
+        double pp_cn = ref_cn[pp_ind * MAX_REF_CN + i];
+        if (pp_cn >= 0.0f) { // Values less that 0 zero are invalid and should not be counted.
+            L_pp[i] = exp(k3 * pp_cn * pp_cn);
+        }
+    }
+    int max_ref_pp = i;
+
+
+    for (int ia = 0; ia < natoms; ia++) { // Loop over all atoms
+
+        const Vec3d pos = Ratoms[ia];
+        const int elem_ind = elems[ia] - 1;
+        const double r_cov_elem = r_cov[elem_ind];
+
+        // Compute the coordination number for this atom
+        double cn = 0;
+        for (int j = 0; j < natoms; j++) {
+            if (j == ia) continue; // No self-interaction for coordination number
+            double d = (Ratoms[j] - pos).norm();
+            double r = r_cov[elems[j] - 1] + r_cov_elem;
+            cn += 1.0f / (1.0f + exp(k12 * r / d + k1));
+        }
+
```

Comment:
Do I understand here that coordication number is calculated from how many other atoms overlap with this atom (using gaussian kernell) ?

Reply:
Yes, it's a measure of how many atoms are nearby. It's not really Gaussian. It's a smooth step function with a radius proportional to the covalent radius of the two atoms, summed over all other atoms.